### ISSUE TITLE
Improve conditional components

### ISSUE THREAD

**Issue by HansOlsson** [modelica/ModelicaSpecification#3489]
The current conditional components are problematic to use due to the restrictions on their use.
We have customers that use them use them extensively (Dymola previously opted for the first solution for 1 below), and find the current rules problematic.

https://specification.modelica.org/master/class-predefined-types-and-declarations.html#conditional-component-declaration

Specifically there are two different sets of issues:
1. The use of a a conditionally enabled component is prohibitively restricted:
   1. Even if we can prove that the use is conditional under the same (or stricter) conditions for its existence it is still illegal. The example in the specification even contains one such example.
   2. One cannot "transform" a conditional component to a normal one; code below.
 2. The use of a conditionally disabled component is limited to connect-statements.

Example of 1.i:
```
  Level1 component1(J=J) if level==1 "Conditional component";
  Level2 component2 if level==2 "Conditional component";
  Level3 component3(J=component1.J) if level<2 "Conditional component"; // Illegal use of component1.J
  Level4 component4(J=component1.J) if level==1 "Conditional component"; // Illegal use of component1.J
```
The Level3 case is from the specification. The Level4 case is even simpler (exactly the same condition), and one could even restrict it to a boolean parameter, and use the same boolean parameter.

Example of 1.ii:
```
 model M
    parameter Boolean isActive=false;
    Real a=time if isActive;
end M;
M m(final isActive=true);
Real x=m.a; // Still illegal, even if final modifier
```

I see two different solutions for 1:
- We just remove the restriction on conditionally enabled components, and view the stricter checks as a quality-of-implementation check, similarly as balancing-check for other parameter-values than the ones currently used.
- We allow it under some very restrictive, but useful conditions:
  - A conditional component can be used in components with exactly the same conditions, or in other cases if the condition evaluates to true and is based on modifiers that are all final. (Exactly how to formulate this isn't clear - we don't have this concept of "final" all the way, and equality of conditions may not always be simple. It is also unclear how useful this will be; in particular we also need to consider 2 below.)

The solution for 2 will require more care (it may also seem less urgent), as we cannot just remove parts of equations at will. I see two related alternatives:
- One possibility would be follow C++ and have something like `if constexpr (x) {` (with an evaluable parameter-value condition); and say that conditionally disabled component can be used in such constructs that are not active.
- Another solution would be that it is always allowed in if-statements in equations, and require that their condition in those cases must be an evaluable expression that evaluates to false (if the component is disabled). One issue is that the use can be inside nested if-statements; which makes it unclear which of those if-statements must have an evaluable condition; it that is an issue I think I would prefer if the restriction was for the top-most one.

**Comment by gkurzbach** [comment]
Is it not enough to say that it is an error if, after evaluating evaluable conditions, references to inactive conditional components remain in the simulation model?

**Comment by HansOlsson** [comment]
> Is it not enough to say that it is an error if, after evaluating evaluable conditions, references to inactive conditional components remain in the simulation model?

That would be a solution - the problem is whether the needed evaluation of evaluable conditions is desired and clear.

**Comment by qlambert-pro** [comment]
I like gerd's suggestion a lot.

**Comment by henrikt-ma** [comment]
I guess the key question then would be whether one is allowed to resolve such errors by evaluating more parameters?

**Comment by HansOlsson** [comment]
> I guess the key question then would be whether one is allowed to resolve such errors by evaluating more parameters?

Sort of, but even that question hides two assumptions: which ones were already evaluated, and is it clear which ones need to be evaluated.

**Comment by eshmoylova** [comment]
I am in favor of lifting the restriction and support @gkurzbach's suggestion together with the current handling of connect equation with inactive conditional components. This was also discussed  in #3391.

**Comment by casella** [comment]
I am not really into the consequence of changing the current rules from a compiler development point of view, but I totally agree that the current rules, which restrict the use of conditional components to connect equations and to their binding equations, are way too restrictive and force people to write code that is unnecessarily convoluted and hard to understand.

In principle, I support @gkurzbach's [proposal](https://github.com/modelica/ModelicaSpecification/issues/3489#issuecomment-1960790254), from a modeller's point of view that would be ideal. I'm not sure if there are some hidden pitfalls in it.

**Comment by gkurzbach** [comment]
> I guess the key question then would be whether one is allowed to resolve such errors by evaluating more parameters?

Isn't it true, that parameters referenced in conditions of conditional declared components have to be evaluated anyway? So, if in other condition expressions  the same parameters occur, die conditions can be evaluated, and the invalid equations in the body are removed. 
If there are some other not evaluated parameters in the condition, there is no need to evaluate them. The model is then just invalid and rejected, because there is still a reference to conditional declared component where the condition is false.

> Sort of, but even that question hides two assumptions: which ones were already evaluated, and is it clear which ones need to be evaluated.

The compiler knows exactly which parameters are already evaluated, and also which need to be evaluated, just those that occur in the expressions of the conditional declarations and all its dependencies.

The both changes in the PR I find cumbersome and hard to understand. I would remove them and write instead:
"It is an error if conditionally declared components whose declaration condition is evaluated to false during compilation could potentially be used in the simulation model."

Btw, SimulationX uses this semantics since the beginning, even before conditional components was defined in Modelica.

**Comment by casella** [comment]
> > I guess the key question then would be whether one is allowed to resolve such errors by evaluating more parameters?
> 
> Isn't it true, that parameters referenced in conditions of conditional declared components have to be evaluated anyway? So, if in other condition expressions the same parameters occur, die conditions can be evaluated, and the invalid equations in the body are removed. If there are some other not evaluated parameters in the condition, there is no need to evaluate them. The model is then just invalid and rejected, because there is still a reference to conditional declared component where the condition is false.
> 
> > Sort of, but even that question hides two assumptions: which ones were already evaluated, and is it clear which ones need to be evaluated.
> 
> The compiler knows exactly which parameters are already evaluated, and also which need to be evaluated, just those that occur in the expressions of the conditional declarations and all its dependencies.
> 
> The both changes in the PR I find cumbersome and hard to understand. I would remove them and write instead: "It is an error if conditionally declared components whose declaration condition is evaluated to false during compilation could potentially be used in the simulation model."

Quoting my colleague Uri Shaked: "It is simple, and it works" 😃

**Comment by casella** [comment]
Earlier on, @gkurzbach [wrote](https://github.com/modelica/ModelicaSpecification/issues/3489#issuecomment-1960790254):

> Is it not enough to say that it is an error if, after evaluating evaluable conditions, references to inactive conditional components remain in the simulation model?

to which @HansOlsson replied:

> That would be a solution - the problem is whether the needed evaluation of evaluable conditions is desired and clear.

We are trying to solve a problem for RTE's Dynawo library, which is currently used for real-time management of the French high-voltage transmission grid. A great success story for Modelica. Unfortunately, the way parameters are currently handled in that library is not good. We believe that using conditional components would be the right way to go, but that would require the specification to go alongside @gkurzbach's [suggestion](https://github.com/modelica/ModelicaSpecification/issues/3489#issuecomment-1960790254). 

Unfortunately, the change to the specification of #3556 did not go in this direction, as it only solves issue 1. stated by @HansOlsson, but not issue 2., which is critical in this case. So, I'm afraid I have to re-open this discussion.

Our use case is illustrated by this simple MWE
```modelica
model M
  parameter Boolean use_p = false;
  parameter Real p if use_p;
  Real x;
equation 
  if use_p then
    x = p;
  else
    x = 2;
  end if;
end M;
```
If `use_p = true`, the parameter `p` is needed in the system equations and the user is expected to provide a meaningful value to it via a modifier, so there is no default binding equation for `p`, and if such a binding equation is not provided upon instantiation, it is an error as per[ Section 8.6](https://specification.modelica.org/master/equations.html#initialization-initial-equation-and-initial-algorithm). On the other hand, if `use_p = false`, then `p` is not needed, so it would be very odd if one had to provide a binding for it anyway to make the model valid.

As a modeller, I think the code shown above solves the problem in the most natural and straightforward way: we make `p` a conditional component, so that if `use_p = false` the component is deleted upon instantiation and there is no need to provide any value (or default value) to it, otherwise it is necessary to provide a parameter value via a modifier upon instantiation.

As I understand also by discussing this with @perost, the modification to the specification text of #3556 is not enough to make this model valid. What we really need is something along the lines of what @gkurzbach [suggested](https://github.com/modelica/ModelicaSpecification/issues/3489#issuecomment-1960790254):
- `if use_p` is an evaluable condition
- after evaluating it, the inactive branch of the if-equation can be ignored
- the remaining equations are always well defined and using active components only, so errors are never reported

Notice that we actually have two use cases with RTE. 

In one case, the model is used within a Modelica GUI (e.g. OMEdit), so the parameter can be evaluated at compile time and the inactive branch removed before proceeding to code generation. Of course, in order to change `use_p`, one needs to recompile the model, which may be slightly inconvenient, but is still possible. With both values of `use_p`, the compiler will not find any reference to inactive components in the simulation model, so no errors will be issued.

In the other case, instead, this component is pre-compiled into a (sort of) FMU, so we need to retain the C-code to support both values of `use_p`, which will be set to either true or false at run time. In general, it may be difficult to figure out statically at compile time that the model is always valid without evaluating the parameters that influence the conditional equations; if that is the case, the generated C code should check this at run time. For example, if you compile this model

```modelica
model M2
  parameter Boolean use_p = false;
  parameter Real p if use_p;
  Real x;
equation 
  if use_p then
    x = p;
  else
    x = 2*p;
  end if;
end M2;
```
and then run it after setting `use_p = false`, a run-time error will be generated as soon as the code for `x := 2*p` is executed, since `p` is actually inactive. Of course, more sophisticated tools could also perform an exhaustive analysis at compile time and figure out this model is (potentially) invalid, that's a quality of implementation issue.

Bottom line: can we please re-consider @gkurzbach's suggestion? I understand SimulationX already implements it. OpenModelica also implemented it for a brief period when we disabled checking the previous limitations and before we added checking the new limitations, but could very easily implement this by removing again the new limitations and adding some run-time checks when the evaluable parameters are actually not evaluated at compile time.

From a modeller's point of view, this would enable writing much easier-to-understand code. In particular, avoiding the need of using protected connectors to generate always-defined aliases of conditionally defined variables, which is really, really ugly.

**Comment by casella** [comment]
Also adding @cbu-hw to the loop.

**Comment by HansOlsson** [comment]
Just for clarity, you can currently handle MWE as follows:
```modelica
model M
  parameter Boolean use_p = false;
  parameter Real p if use_p;

  Modelica.Blocks.Sources.RealExpression realExpression_p(y=p) if use_p
    annotation (Placement(transformation(extent={{-60,42},{-40,62}})));
  Modelica.Blocks.Sources.RealExpression realExpression_2(y=2) if not use_p
    annotation (Placement(transformation(extent={{-58,-40},{-38,-20}})));
  Modelica.Blocks.Interfaces.RealOutput x
    annotation (Placement(transformation(extent={{98,4},{118,24}})));
equation 

  connect(realExpression_p.y, x) annotation (Line(points={{-39,52},{90,52},{90,14},
          {108,14}}, color={0,0,127}));
  connect(realExpression_2.y, x) annotation (Line(points={{-37,-30},{90,-30},{90,
          14},{108,14}}, color={0,0,127}));
  annotation (uses(Modelica(version="4.1.0")));
end M;
```

**Comment by HansOlsson** [comment]
Also discussed `if constexpr` in C++, https://en.cppreference.com/w/cpp/language/if.html#Constexpr_if

**Comment by HansOlsson** [comment]
More complicated model discussed at meeting:
```modelica
model M
  parameter Boolean use_p = false;
  parameter Boolean use_p2 = true if not use_p ;
  parameter Real p if use_p;
  Real x;
  Boolean y1 = use_p and p>0;
  Boolean y2 = p>0 and use_p;
  // Henrik (similar to other rules for if-then-else)
  Boolean y3 = if use_p then p>0 else false;
equation 
  // if use_p or someComplicatedExpression(...) then
  if use_p and not use_p2 then
    x = p;
  else
    x=2;
  end if;
end M;
```

**Comment by HansOlsson** [comment]
Seems too complicated to resolve for 3.7

**Comment by casella** [comment]
> Just for clarity, you can currently handle MWE as follows:
> 
> model M
>   parameter Boolean use_p = false;
>   parameter Real p if use_p;
> 
>   Modelica.Blocks.Sources.RealExpression realExpression_p(y=p) if use_p
>     annotation (Placement(transformation(extent={{-60,42},{-40,62}})));
>   Modelica.Blocks.Sources.RealExpression realExpression_2(y=2) if not use_p
>     annotation (Placement(transformation(extent={{-58,-40},{-38,-20}})));
>   Modelica.Blocks.Interfaces.RealOutput x
>     annotation (Placement(transformation(extent={{98,4},{118,24}})));
> equation 
> 
>   connect(realExpression_p.y, x) annotation (Line(points={{-39,52},{90,52},{90,14},
>           {108,14}}, color={0,0,127}));
>   connect(realExpression_2.y, x) annotation (Line(points={{-37,-30},{90,-30},{90,
>           14},{108,14}}, color={0,0,127}));
>   annotation (uses(Modelica(version="4.1.0")));
> end M;

Sure. It's just horrible and completly obscure, if compared to the proposed code.

**Comment by casella** [comment]
> Seems too complicated to resolve for 3.7

OK, I agree that handling the general case is tricky and maybe unnecessarily complicated. Here is my proposal: amend the text in [Section 4.4.5](https://specification.modelica.org/master/class-predefined-types-and-declarations.html#conditional-component-declaration) as follows (modifications are in boldface)

> If the Boolean expression is false, the component (including its modifier) is removed from the flattened DAE, and connections to/from the component are removed.
> Such a component can only be modified, used in connections, used in a modifier of another conditional component with a false condition, **and/or used in a branch of an if-equation with an evaluable false condition.**
> There are no restrictions on the component if the Boolean expression is true.

This would make [my MWE](https://github.com/modelica/ModelicaSpecification/issues/3489#issuecomment-4246144382) legal. All the considerations I made in that comment remain valid.

As I understand, the implementation of this amended rule is trivial: if the condition of the if-equation is evaluated at compile time, only one branch remains and then it can be immediatly verified if there are references to inactive component in the surviving branch, in which case an compile-time error is issued. Otherwise, this check has to be postponed to run-time, when the condition is actualy evaluated, possibly generating a run-time error. @perost would you like to comment on that?

What do you think? Can we have this in 3.7?

**Comment by henrikt-ma** [comment]
No, as the plan for wrapping up 3.7 is to only include things which we can wrap up quickly.  We talked about this PR for hours during the meeting, and since there is a long history of not finding the "right" semantics for conditional components it was concluded that we shouldn't rush it this time.

What hasn't been communicated yet, however, is that the current plan is to release 3.8 once we have a first version of unit checking in place.  It doesn't seem infeasible to work out a new design for conditional components in parallel, so you might not have to wait that long for a resolution in a released version after all.

**Comment by casella** [comment]
No need to rush. On the other hand, I would warmly recommend to seek for incremental and practical solutions, rather than aiming at very general grand schemes that may take forever to materialize.

I guess the reason why it's been hard to find the right semantics of conditional components is that a completely general solution may be elusive. On the other hand, what we had for 25+ years and still have now is so ugly.

So, please do consider more restricted approaches, like the extension I suggested. It's always possible to make them more general in the future.
